### ISSUE TITLE
Sonos binding: NPE when using player item for a slave Sonos

### ISSUE THREAD

**Issue by lolodomo** [eclipse/smarthome#1902]
When using the player (play, pause, next, prev buttons) with UI for an item associated to a Sonos being only a slave member of a group, it triggers a null pointer exception:
```

2016-07-23 13:28:47.975 [ERROR] [ome.core.thing.internal.ThingManager] - Exception occured while calling handler: java.lang.NullPointerException
java.util.concurrent.ExecutionException: java.lang.NullPointerException
        at java.util.concurrent.FutureTask.report(FutureTask.java:122)[:1.8.0]
        at java.util.concurrent.FutureTask.get(FutureTask.java:206)[:1.8.0]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.callAsynchronous(SafeMethodCaller.java:179)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:72)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at org.eclipse.smarthome.core.common.SafeMethodCaller.call(SafeMethodCaller.java:56)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at org.eclipse.smarthome.core.thing.internal.ThingManager.receiveCommand(ThingManager.java:368)[104:org.eclipse.smarthome.core.thing:0.8.0.b5]
        at org.eclipse.smarthome.core.items.events.AbstractItemEventSubscriber.receive(AbstractItemEventSubscriber.java:46)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:192)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at org.eclipse.smarthome.core.internal.events.OSGiEventManager$1.call(OSGiEventManager.java:1)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at org.eclipse.smarthome.core.common.SafeMethodCaller$CallableWrapper.call(SafeMethodCaller.java:170)[98:org.eclipse.smarthome.core:0.8.0.b5]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)[:1.8.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)[:1.8.0]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)[:1.8.0]
        at java.lang.Thread.run(Thread.java:744)[:1.8.0]
Caused by: java.lang.NullPointerException
        at org.eclipse.smarthome.binding.sonos.handler.ZonePlayerHandler.handleCommand(ZonePlayerHandler.java:292)[213:org.eclipse.smarthome.binding.sonos:0.8.0.b5]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$2.call(ThingManager.java:371)[104:org.eclipse.smarthome.core.thing:0.8.0.b5]
        at org.eclipse.smarthome.core.thing.internal.ThingManager$2.call(ThingManager.java:1)[104:org.eclipse.smarthome.core.thing:0.8.0.b5]
        ... 5 more
2016-07-23 13:28:47.986 [INFO ] [smarthome.event.ItemCommandEvent    ] - Item 'SonosSalonControl' received command NEXT
```

Tested with beta 3.

**Comment by lolodomo** [comment]
I investiguated this one too.
The bug is in method getHandlerByName. It only searches for things having the same type as the current thing. But my Sonos are of different types sonos:CONNECTAMP and sonos:PLAY5.
And there is no reason to put the second loop inside the first.
I will propose a fix.

**Comment by lolodomo** [comment]
And there is the same error in restoreAllPlayerState and saveAllPlayerState.

**Comment by kgoderis** [comment]
Initially there was only one type of Sonos thing but that got changed some time ago by someone
I do not recall why I had that loop there, there must have been a reason so better test it afterwards so that we do not break other stuff

**Comment by lolodomo** [comment]
This loop is to get the handler of the group coordinator.
In fact, there are first one search considering the default UID that would be given by the automatic discovery and then a second search through all things.

**Comment by R-Edelhoff** [comment]
The method getHandlerByName() is supposed to return the ThingHandler instance for the given Sonos player UDN (RinconID), which is used to invoke actions on another player then the currently handled one. This is required for delegating commands to the coordinator speaker when invoking basic control actions on a slave. 
According to your stacktrace you invoke the NEXT command which fetches the coordinator ThingHandler and invokes the method there. It seems that there is no ThingHandler instance available for the coordinator speaker of your Sonos group configuration so that we get a NPE.

Have you paired both devices or just the PLAY5?

**Comment by lolodomo** [comment]
As previously explained there was a bug in this method that only consider things of the same type of the current thing. My 2 Sonos are paired to things but are of different types.
This problem is fixed by my PR.

NPE were possible if the group coordinator was not paired to a thing (not my case). This is fixed in my PR too.

**Comment by R-Edelhoff** [comment]
Regarding restoreAllPlayerState and saveAllPlayerState you are right. It will only be invoked on the currently handled ThingHandler instance. It should be invoked for all registered ThingHandlers of different player types.

**Comment by kgoderis** [comment]
I initially had the comparison in there just to speed up the loop, but that was before the proliferation of thing types. We should scan the remainder of the .java for similar bits of code. 

If I remember well the getHandlerByName() can take either a rincon id or a thing ID, so the two loops need to stay in there. The type comparison can indeed be removed, unless we want to comparie on the binding ID.... In an ecosystem with a lot of things this might have a performance impact

**Comment by lolodomo** [comment]
I kept the two loops but suppressed their imbrication. First loop is unchanged. Second loop was fixed but still checks the thing type.

**Comment by R-Edelhoff** [comment]
The bug fixes done within this PR are fine in my opinon. 

However i would not introduce a nullcheck everytime we request the coordinatorHandler.
Reporting a warning in this case is good. But additionally i would throw an unchecked runtime exception when the coordinator is not available (paired) to indicate that the system will not behave as expected.
The outcome in both ways (nullchecks or exception) is anyway the same... the command will not be executed.
@kgoderis What do you think?

**Comment by kgoderis** [comment]
Fine for me, is also what was suggested earlier on

**Comment by lolodomo** [comment]
I need a little time to finidh the change and test it. Should be done probably before the end of this week.

**Comment by kaikreuzer** [comment]
fixed by #1905
